docs/architecture/adr-050-improved-trusted-peering.md

```diff
@@ -0,0 +1,60 @@
+# ADR 50: Improved Trusted Peering
+
+## Changelog
+* 22-10-2019: Initial draft
+* 05-11-2019: Modify `maximum_dial_period` to `persistent_peers_max_dial_period`
+
+## Context
+
+When `max_num_inbound_peers` or `max_num_outbound_peers` of a node reached, the node cannot spare more slots to any peer 
+by inbound or outbound. Therefore, after certain period of disconnection, any important peering can be lost indefinitely 
+because all slots consumed by other peers, and the node stop trying to dial the peer anymore.
+
+This is happening because of two reasons, exponential backoff and absence of unconditional peering feature for trusted peers.
+
+
+## Decision
+
+We would like to suggest solving this problem by introducing two parameters in `config.toml`, `wildcard_peer_ids` and 
+`persistent_peers_max_dial_period`. 
+
+1) `wildcard_peer_ids`
+
+A node operator inputs list of ids of peers which are allowed to be connected by both inbound or outbound regardless of 
+`max_num_inbound_peers` or `max_num_outbound_peers` of user's node reached or not.
```

Comment:
Are there limits on the number of trusted/wildcard peers a node can specify? And does adding a trusted/wildcard peer effectively increase the maximum number of peers by 1 (e.g. is the maximum number of inbound peers specified by `max_num_inbound_peers` + n, where n is the number of wildcard peers)?

Reply:
1. There is no limits for wildcard peers. I think it is OK without the limit. Do we need it?

2. The peers in wildcard_peer_ids is not objective to `max_num_inbound_peers`. So we don't add `n`. The state machine just does not check `max_num_inbound_peers` criteria for any wildcard_peer_ids. Also the state machine just does not count any peer in wildcard_peer_ids to compare `max_num_inbound_peers`

Reply:
NOTE: we will have a problem with "wildcard_peer_ids is not objective to max_num_inbound_peers" because there's usually a limited number of connections per OS process (`ulimit`)

Reply:
It sounds like we should have some kind of limit here.

Reply:
I think the OS limit is much far larger than what we need to connect practically. 200 peers are what we practically think maximum, and it is nowhere close to default limit on OS.

Also users can aware that maxpeernum + numofunconditionalpeers = actual possible number of peers. So it is not a unpredictable risk to manage.

For such reason, I think it can be stated in ADR, but we dont think it is needed to be implemented in this spec.